<a href="https://colab.research.google.com/github/manyamanikandan447/Regression-assignment/blob/main/regression_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#  Regression Assignment – California Housing Dataset

In [ ]:
from sklearn.datasets import fetch_california_housing
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Load dataset
housing = fetch_california_housing(as_frame=True)
df = housing.frame

# Check for missing values
print(df.isnull().sum())

# Features and target
X = df.drop('MedHouseVal', axis=1)
y = df['MedHouseVal']

# Standardization
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

MedInc         0
HouseAge       0
AveRooms       0
AveBedrms      0
Population     0
AveOccup       0
Latitude       0
Longitude      0
MedHouseVal    0
dtype: int64


### Preprocessing
- Converted dataset into a Pandas DataFrame.  
- Checked for missing values (none found).  
- Standardized features using `StandardScaler` because models like SVR and Gradient Boosting are sensitive to scale.

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR

models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(random_state=42),
    "Random Forest": RandomForestRegressor(random_state=42),
    "Gradient Boosting": GradientBoostingRegressor(random_state=42),
    "SVR": SVR()
}

### Algorithms
- **Linear Regression:** Baseline model, fits a straight line.  
- **Decision Tree:** Splits data into regions, captures non-linearities.  
- **Random Forest:** Ensemble of trees, reduces overfitting.  
- **Gradient Boosting:** Sequential trees correcting errors, strong performance.  
- **SVR:** Kernel-based, sensitive to scaling.

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

results = {}
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    results[name] = {
        "MSE": mean_squared_error(y_test, y_pred),
        "MAE": mean_absolute_error(y_test, y_pred),
        "R2": r2_score(y_test, y_pred)
    }

results_df = pd.DataFrame(results).T
results_df

,MSE,MAE,R2
Linear Regression,0.555892,0.533200,0.575788
Decision Tree,0.494272,0.453784,0.622811
Random Forest,0.255498,0.327613,0.805024
Gradient Boosting,0.293999,0.371650,0.775643
SVR,0.355198,0.397763,0.728941


### Results

| Model              | MSE  | MAE  | R²  |
|--------------------|------|------|-----|
| Linear Regression  | 0.55 | 0.53 | 0.57 |
| Decision Tree      | 0.49 | 0.45 | 0.62 |
| Random Forest      | 0.25 | 0.32 | 0.80 |
| Gradient Boosting  | 0.29 | 0.37 | 0.77 |
| SVR                | 0.35 | 0.39 | 0.72 |

### Insights
- **Best-performing:** Random Forest Regressor(lowest error, highest R²).   
- **Worst-performing:** Linear Regression, limited by linear assumptions.